src/app/clusters/operational-state-server/operational-state-server.cpp

```diff
@@ -0,0 +1,417 @@
+        // even if errors happen after this. This ensures that we don't execute any fall-back strategies
+        // to handle this command since at this point, the caller is taking responsibility for handling
+        // the command in its entirety, warts and all.
+        //
+        handlerContext.SetCommandHandled();
+
+        if (DataModel::Decode(handlerContext.mPayload, requestPayload) != CHIP_NO_ERROR)
+        {
+            handlerContext.mCommandHandler.AddStatus(handlerContext.mRequestPath,
+                                                     Protocols::InteractionModel::Status::InvalidCommand);
+            return;
+        }
+
+        func(handlerContext, requestPayload);
+    }
+}
+
+CHIP_ERROR OperationalStateServer::GetOperationalStateList(OperationalStateStructDynamicList ** operationalStateList, size_t & size)
+{
+    return mOperationalStateDataProvider.LoadOperationalStateList(endpointId, clusterId, operationalStateList, size);
+}
+
+void OperationalStateServer::ReleaseOperationalStateList(OperationalStateStructDynamicList * operationalStateList)
+{
+    mOperationalStateDataProvider.ReleaseOperationalStateList(operationalStateList);
+}
+
+CHIP_ERROR OperationalStateServer::ClearOperationalStateList()
+{
+    return mOperationalStateDataProvider.ClearOperationalStateList(endpointId, clusterId);
+}
+
+CHIP_ERROR OperationalStateServer::SetPhaseList(const PhaseList & phaseList)
+{
+    return mOperationalStateDataProvider.StorePhaseList(endpointId, clusterId, phaseList);
+}
+
+CHIP_ERROR OperationalStateServer::GetPhaseList(PhaseListCharSpan ** phaseList, size_t & size)
+{
+    return mOperationalStateDataProvider.LoadPhaseList(endpointId, clusterId, phaseList, size);
+}
+
+void OperationalStateServer::ReleasePhaseList(PhaseListCharSpan * phaseList)
+{
+    mOperationalStateDataProvider.ReleasePhaseList(phaseList);
+}
+
+CHIP_ERROR OperationalStateServer::ClearPhaseStateList()
+{
+    return mOperationalStateDataProvider.ClearPhaseStateList(endpointId, clusterId);
+}
+
+CHIP_ERROR OperationalStateServer::SetOperationalState(OperationalStateStruct & opState)
+{
+    OperationalStateStructDynamicList op;
+
+    op.operationalStateID = static_cast<Clusters::OperationalState::OperationalStateEnum>(opState.OperationalStateID);
+
+    ReturnErrorOnFailure(mOperationalStateDataProvider.UseOpStateIDGetOpStateStruct(endpointId, clusterId, op));
+
+    if (&operationalState != &opState)
+    {
+        char * dest = const_cast<char *>(operationalState.OperationalStateLabel);
+        size_t len  = op.operationalStateLabel.size();
+        memset(dest, 0, sizeof(operationalState.OperationalStateLabel));
+        memcpy(dest, op.operationalStateLabel.data(), len);
+        operationalState.OperationalStateID = static_cast<uint8_t>(op.operationalStateID);
+    }
+    return CHIP_NO_ERROR;
+}
+
+OperationalStateStruct & OperationalStateServer::GetOperationalState()
```

Comment:
Why is this returning a non-const reference?

Reply:
Good suggestion. Done.